Oliver/issue 365 travel times backend

Oliver-Zen opened the pull request:
### Summary

Resolves #365

<!-- What does this PR change and why? Discuss any breaking changes. -->

### Checklist

- [x] Follow the first two steps from #364 to get a matrix of distances between different class locations we already have. You can also coordinate with whoever is doing that ticket to just get the matrix from them.
- [x] Create a utility functions in `src/utils/mapbox` which given two locations, finds the distances between them either using the existing matrix or the [Mapbox Matrix API](https://docs.mapbox.com/api/navigation/matrix/) if the existing location isn't there in the matrix
- [x] In the `MapView` component, add a function that takes the given locations and using the utility functions you made and gets the distances between them. Make sure you batch the calls together to prevent increased latency. Use `useMemo` to avoid repeating the computation each time.


### Notes
- Distance matrix generated by [Mapbox Matrix API](https://docs.mapbox.com/api/navigation/matrix/)
  - Currently there are 62 GT locations but with 9 duplicated locations (same latitude and longitude), so distance matrix is 53 * 53
- Utility functions use matrix first with fallback using [Mapbox Directions API](https://docs.mapbox.com/api/navigation/directions/)

### How to Test
- To test Mapbox utility functions:
```bash
npm test src/utils/mapbox/travelTimes.test.ts
```
- To test distance matrix generation:
```zsh
export REACT_APP_MAPBOX_TOKEN=mapbox_token_with_matrix_downloading_scope
node src/utils/mapbox/scripts/populate-distance-matrix.js 
```
- To test component, use React dev tool:
<p align="center">
  <img width="500" height="444" alt="image" src="https://github.com/user-attachments/assets/0e8646fb-b417-4101-9e5d-618040d299ac" /><br>
  <em> travelTimes state shown in Tuesday view</em>
</p>
<p align="center">
  <img width="500" height="450" alt="image" src="https://github.com/user-attachments/assets/9a6c7d93-520a-4cd1-8e40-f5023a0fc143" /><br>
  <em> travelTimes state is null in All Days view</em>
</p>

Oliver-Zen committed:
feat: add GT campus locations and distance matrix generation

Oliver-Zen committed:
feat: add GT campus locations and distance matrix generation

Oliver-Zen committed:
feat: implement travel time calculation utilities

Oliver-Zen committed:
feat: implement travel time calculation utilities

Oliver-Zen committed:
feat: integrate travel time backend into MapView component

Oliver-Zen committed:
feat: integrate travel time backend into MapView component

Oliver-Zen committed:
test: mock Mapbox API calls to fix CI failures

Oliver-Zen committed:
test: mock Mapbox API calls to fix CI failures

Oliver-Zen committed:
fix: resolve ESLint errors for CI build

Oliver-Zen committed:
fix: resolve ESLint errors for CI build

uma-anand commented on src/components/MapView/index.tsx:
```diff
@@ -29,6 +32,42 @@ export default function MapView({
     zoom: 15,
   });
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
is this still needed?

oliverxt replied:
since `travelTimes` is declared but never used, I think we need this comment.

and I think we can't have `setTravelTimes` without `travelTimes` since useState returns both the state and setter as a pair.

uma-anand replied:
Makes sense. Sorry, forgot we're not actually displaying it right now.

uma-anand commented on src/components/MapView/index.tsx:
```diff
@@ -17,10 +18,12 @@ export type MapLocation = {
 
 export type MapViewProps = {
   locations: MapLocation[];
+  activeDay?: string;
```
Instead of making MapViewProps take in activeDay, make it a flag here that is hasTravelTimes with a true or a false and pass this from the higher function. This compoenent shld be independent of day.

oliverxt replied:
resolved with new commit~

uma-anand commented on src/utils/mapbox/travelTimes.test.ts:
thanks for writing a test!

uma-anand reviewed (changes requested):
Code is super super clean!! Excited for this feature. There is going to be a lot of weird merge logic once we get the recurring events locations in but for now looks good! Left a nit or two.

Oliver-Zen committed:
refactor: better prop in MapView component

Oliver-Zen committed:
refactor: better prop in MapView component

uma-anand reviewed (approved):
LGTM! Thanks for working on this. Great job!

Oliver-Zen committed:
feat(map): add travel distance overlay

Oliver-Zen committed:
feat(map): add travel distance overlay

Oliver-Zen commented:
## Test

### ALL Day

<img width="843" height="341" alt="image" src="https://github.com/user-attachments/assets/09d8d4a8-a2e7-4094-9abd-88fc3eb126ec" />

- Can't view travel times in ALL day

### Weekday

<img width="834" height="316" alt="image" src="https://github.com/user-attachments/assets/546936d2-0fd6-4561-986e-4b22bff692b1" />

<img width="842" height="320" alt="image" src="https://github.com/user-attachments/assets/a00659da-450d-49cf-98b4-27eb18bbbb4c" />

- Used GT branch color "[tech gold](https://brand.gatech.edu/our-look/colors)"
- Chose solid no-arrow line instead of dash arrowed line considering this situation: location A --> location B --> location A
  - A solid no-arrow line is more elegant (to me)

uma-anand commented on src/components/Map/stylesheet.scss:
```diff
@@ -8,3 +8,87 @@
+.travel-toggle {
+  border-radius: 999px;
+  border: none;
+  background-color: rgba(15, 23, 42, 0.85);
```
For all colors, we store in the variables.scss. First check if that color already exists and if it doesn't, add a new variable.

uma-anand commented on src/components/Map/index.tsx:
```diff
@@ -80,7 +94,24 @@ export default function Map(): React.ReactElement {
       />
-      <MapView locations={activeLocations} />
+      <div className="map-view-panel">
+        <button
```
Make this a new component "ToggleButton"

uma-anand reviewed (changes requested):
Good work on getting the arrows to appear!!

We want to follow the Figma design exactly. Can you make sure we have the correct colors, dashing, arrows, and icons as the design? For both the map and the button. For the button, make sure light mode is implemented properly.

Also, can you rebase onto sprint-4? In Sprint-4 we took a lot of the Map stuff out into its own file. Also you're going to want to check behavior with recurring events.

Oliver-Zen committed:
fix(lint): ensure travel line style hook runs consistently before token check

Oliver-Zen committed:
refactor(map): update travel times to use Figma design specs

to squash

wip

Oliver-Zen committed:
refactor(map): extract ToggleButton into component

Oliver-Zen committed:
feat(map): add light mode for travel times ToggleButton

Oliver-Zen commented:
### PR Update Summary
- rebased on sprint 4
- aligned with Figma design
- extracted ToggleButton

### Test
#### ALL day
<img width="1582" height="1034" alt="image" src="https://github.com/user-attachments/assets/79dcf08e-e76e-4088-b557-f23097c03b52" />

#### Light mode, Course + Event
<img width="1582" height="1034" alt="image" src="https://github.com/user-attachments/assets/6c8a96db-1724-4732-81e8-9929cc278ad4" />
<img width="1582" height="1034" alt="image" src="https://github.com/user-attachments/assets/b8d6d1f7-db2a-4b3e-8433-04fb91d0c8e8" />

#### Dark mode, cyclic travel
<img width="1582" height="1034" alt="image" src="https://github.com/user-attachments/assets/e526bee5-ca0a-4437-acb6-3feac352ea13" />
<img width="1582" height="1034" alt="image" src="https://github.com/user-attachments/assets/4ce90985-cf95-441a-919b-a9b96be7ebd0" />

uma-anand committed:
finish merge

uma-anand committed:
fix header positioning relative to toggle button
